### ISSUE TITLE
Picklist - Emit `onMove...()` events before source/target lists are modified

### ISSUE THREAD

**Issue by marshalhayes** [primefaces/primeng#12028]
### Describe the feature you would like to see added

When using the picklist component, we can tie into `onMoveToTarget()` and `onMoveAllToTarget()` as well as `onMoveToSource()` and `onMoveAllToSource()` to listen for when items are moved.

All of those events emit the items that were moved as `$event.items`. 

I was trying to find a way to restrict certain items in the target list from being removed. So, like this:

```js
onMoveToSource($event) {
  // When items are moved from target to source, prevent required items from being moved
  $event.items = $event.items.filter(item => !item.required);
}
```

My thinking was that, if I modify `$event.items` (so in this case, removing `required` items), that those would be the ones moved to the source. However, this doesn't currently seem possible. Peeking at the source for the picklist component, it looks like the `onMove...()` events are emitted after the source/target arrays have already been modified.

Could the `onMove...()` events be emitted before the source/target arrays are modified?

**Comment by lixaotec** [comment]
I agree with the need.. all components should have a way to let user deal with the event without default behavior.

So we can run validations, cancel an event for some business rules
